### PR TITLE
[CI]: Kuberay operator e2e tests

### PR DESCRIPTION
**Pull request by astefanutti**
## Why are these changes needed?

This PR adds e2e tests for the operator, following up https://github.com/ray-project/kuberay/pull/1539#issuecomment-1778717122.

This e2e testing enables to assert Kuberay resources, in a way that cannot be achieved with existing tests (envtest).

The e2e tests execution output is post-processed using gotestfmt:

image

The logs and events are gathered and uploaded are part of the workflow run:

image

### CODE REVIEW

**Comment by kevin85421** [comment]
E2E tests are great, but I prefer unit tests and `envtest`. However, considering we already have some E2E tests in place, maintaining multiple types of E2E tests could become burdensome. cc @hongchaodeng @architkulkarni

**Comment by astefanutti** [comment]
> E2E tests are great, but I prefer unit tests and `envtest`. However, considering we already have some E2E tests in place, maintaining multiple types of E2E tests could become burdensome. cc @hongchaodeng @architkulkarni

I'd argue implementing unit tests to cover the issues mentioned in https://github.com/ray-project/kuberay/pull/1539#issuecomment-1778717122 is very difficult, or cumbersome.

For the existing e2e tests, I'd be inclined to propose we migrate them over to Golang, so they are strongly typed, and benefit from the Kubernetes ecosystem, following the example in this PR.

**Review by architkulkarni — COMMENTED** [review]
Thanks for the contribution, I'll review this soon. One initial comment, I remember we couldn't use github actions for e2e tests because the github actions free tier has a CPU limit, which prevented the number of Ray pods we could run.  To get around this, we added a Buildkite CI runner which didn't have this limit.  Is it feasible to move the test to Buildkite?

(No need to get started on the implementation if the envtest vs golang question isn't resolved yet. Personally, I don't have a strong preference here and I'm okay with migrating to golang in the long term)

**Comment by astefanutti** [comment]
> Thanks for the contribution, I'll review this soon. One initial comment, I remember we couldn't use github actions for e2e tests because the github actions free tier has a CPU limit, which prevented the number of Ray pods we could run. To get around this, we added a Buildkite CI runner which didn't have this limit. Is it feasible to move the test to Buildkite?

I'm not familiar with Buildkite, but I don't see any reasons why these tests could not be run with it.

**Review by blublinsky — APPROVED** [review]
LGTM

File: ray-operator/Makefile
Comment (z103cb) [inline ray-operator/Makefile:66]: can WHAT variable be initialised with a `?=`

**Review by z103cb — COMMENTED** [review]
Would it make sense to have a `e2e-test` target in the operator make? (I prefer to run most thing locally, before pushing to GitHub)

File: ray-operator/test/support/environment.go
Comment (z103cb) [inline ray-operator/test/support/environment.go:23]: I am not sure about how you feel about adding logic to support arm64 images. IMHO, using the `-aarch64` should be the case if we detect that the host the test is run on has that architecture. I have simillar code in the apiserver e2e test (PR)

**Review by z103cb — APPROVED** [review]
Overall this is looking good, left a couple of questions (nitpicks)

**Comment by z103cb** [comment]
> E2E tests are great, but I prefer unit tests and `envtest`. However, considering we already have some E2E tests in place, maintaining multiple types of E2E tests could become burdensome. cc @hongchaodeng @architkulkarni  

I woud agree with @astefanutti,  unit tests are good, particularly if they focus on "unhappy paths".  

In my experience having an extensive setup of mock objects adds needless complexity, and most importantly it  will provide a false sense of confidence.  Being able to test the operator behavior on a small cluster (kind) it is better as it allows for catching behaviours that otherwise you's not think about.

**Comment by kevin85421** [comment]
Sync with @architkulkarni offline. He has agreed to maintain the new test framework for RayJob, so I am fine with a new test framework.

> For the existing e2e tests, I'd be inclined to propose we migrate them over to Golang, so they are strongly typed, and benefit from the Kubernetes ecosystem, following the example in this PR.

We can discuss this if we observe enough benefits from this new test framework. Currently, most contributors are comfortable with the Python e2e tests.

> E2E tests are great, but I prefer unit tests and envtest.

I found my statement is not clear enough. My point is not about choosing between unit tests/envtest and e2e tests. We should have both.

**Review by architkulkarni — APPROVED** [review]
Looks good to me!

File: .github/workflows/actions/kind/action.yml
Comment (architkulkarni) [inline .github/workflows/actions/kind/action.yml:51]: What's the purpose of the ingress controller?

File: ray-operator/test/e2e/fail.py
Comment (architkulkarni) [inline ray-operator/test/e2e/fail.py:3]:
```suggestion
print("Something is seriously wrong.", file=sys.stderr)
```
I think we want to use Python 3 here? Though I guess it doesn't matter because we want it to fail anyway

**Review by architkulkarni — APPROVED** [review]
@astefanutti Thanks for the contribution! I'm okay with merging this, once the open questions/comments on the PR are answered. I've also added some minor comments/questions.

I think some kind of brief README explaining the test architecture would go a long way towards helping future maintainers.

File: ray-operator/test/e2e/fail.py
Comment (astefanutti) [reply ray-operator/test/e2e/fail.py]: Agreed and updated.

File: .github/workflows/actions/kind/action.yml
Comment (astefanutti) [reply .github/workflows/actions/kind/action.yml:51]: It's not directly used in this PR, but it'll be required when test cases that cover ingresses to RayService or Ray cluster dashboards will be added.

File: ray-operator/Makefile
Comment (astefanutti) [reply ray-operator/Makefile]: Done, PTAL.

**Comment by astefanutti** [comment]
> I think some kind of brief README explaining the test architecture would go a long way towards helping future maintainers.

@architkulkarni I've update the `DEVELOPMENT.md` to document how to run these new e2e tests. This file is already quite large, so I wonder where would documenting the architecture would be best suited. Any suggestions?

File: ray-operator/test/support/environment.go
Comment (astefanutti) [reply ray-operator/test/support/environment.go:23]: While it would solve the problem locally, I think this should be addressed holistically by publishing multi-architecture container images, with the tags pointing to the multi-architecture manifests, so the right image is resolved by the container runtime when pulling the image. I've created #1630 to track that.

**Comment by astefanutti** [comment]
> Would it make sense to have a `e2e-test` target in the operator make? (I prefer to run most thing locally, before pushing to GitHub)

@z103cb I've added the `test-e2e` target in the operator Makefile, PTAL.

**Review by z103cb — APPROVED** [review]

File: ray-operator/DEVELOPMENT.md
Comment (blublinsky) [inline ray-operator/DEVELOPMENT.md:62]: Do you want to load nightly? It might not have all your current development. I think its better to build the image locally and then load it. See how API server make does it.

File: ray-operator/test/e2e/counter.py
Comment (blublinsky) [inline ray-operator/test/e2e/counter.py:1]: There are already a few versions of this code in the project. See for example https://github.com/ray-project/kuberay/pull/1460/files#diff-bc35e167b5138f35311d7d47ad10f7439d8e31404c47fc09baec1e74a87e95d1.
Can we have 1 version of it?

File: ray-operator/test/e2e/rayjob_cluster_selector_test.go
Comment (blublinsky) [inline ray-operator/test/e2e/rayjob_cluster_selector_test.go:45]: Can we reuse definitions from the unit tests? We seem to repeat these definitions all over the place.

File: ray-operator/test/e2e/rayjob_cluster_selector_test.go
Comment (blublinsky) [inline ray-operator/test/e2e/rayjob_cluster_selector_test.go:169]: Same as above. Should we create a set of definitions and use it both in unit tests and here

**Review by blublinsky — APPROVED** [review]
There are several nitpicks, but overall it looks good

File: ray-operator/DEVELOPMENT.md
Comment (astefanutti) [reply ray-operator/DEVELOPMENT.md:62]: While the image reference is `kuberay/operator:nightly`, it's actually built from the local source code in the previous step, and then loaded into KinD. So despite the reference being a bit confusing, I think it does take current development into account.

File: ray-operator/test/e2e/rayjob_cluster_selector_test.go
Comment (astefanutti) [reply ray-operator/test/e2e/rayjob_cluster_selector_test.go:169]: These are not strictly identical to the objects used in the unit tests. There might be some opportunities to factorise things, but I'd be inclined to do it once we have apply configurations generated for the APIs, which I plan to do add ASAP.

File: ray-operator/test/e2e/counter.py
Comment (astefanutti) [reply ray-operator/test/e2e/counter.py:1]: What about updating the API server e2e tests to use that file from the operator e2e tests once this is merged? I'd rather have the API server module "depending" on the operator module, rather than the opposite.

Comment (blublinsky) [reply ray-operator/test/e2e/counter.py:1]: Between you and Laurentiu

File: ray-operator/test/e2e/rayjob_cluster_selector_test.go
Comment (blublinsky) [reply ray-operator/test/e2e/rayjob_cluster_selector_test.go:169]: Sounds good

Comment (blublinsky) [reply ray-operator/test/e2e/rayjob_cluster_selector_test.go:169]: Sounds good

**Pull request merged**
